[WTF] Add LazyRef & LazyUniqueRef

Constellation opened the pull request:
#### 339bfd3e4a54c1e215bc300530670cfb2350b04d
<pre>
[WTF] Add LazyRef &amp; LazyUniqueRef
<a href="https://bugs.webkit.org/show_bug.cgi?id=267830">https://bugs.webkit.org/show_bug.cgi?id=267830</a>
<a href="https://rdar.apple.com/121328458">rdar://121328458</a>

Reviewed by Ryosuke Niwa.

This patch adds LazyRef and LazyUniqueRef. This is similar to LazyProperty in JSC.
We can set *stateless* lambda in the constructor side of the owner object so that it offers clean interface for lazy initialization,
which does not mess up the owner object&apos;s interface.
For example,

    LazyUniqueRef&lt;VM, Property&gt; m_property;
    Property&amp; property() { return m_property.get(*this); }

First parameter is owner, which needs to be passed to `get`. And then, this `get` will automatically initialize if it is not initialized.
And the initialization code can be freely customized, but you do not need to define it as a VM&apos;s member! Instead, you can define it as a lambda inside VM.cpp.

    // This lambda needs to be stateless. So it must not capture anything (otherwise, it hits crash because of RELEASE_ASSERT anyway).
    m_property.initLater(
        [](VM&amp; vm, auto&amp; ref) {
            // You can do whatever. Even you can invoke the other property which can be further lazily initialized (so, it implicitly creates dependency graph of initialization).
            ref.set(Property::create());
            // And further, you can do anything after setting it. So these operations can see Property&amp; via `vm.property()`.
        });

Or you can set lambda in constructor, so like, VM&apos;s constructor list,

    : m_property(
        [](VM&amp; vm, auto&amp; ref) {
            // You can do whatever. Even you can invoke the other property which can be further lazily initialized (so, it implicitly creates dependency graph of initialization).
            ref.set(Property::create());
            // And further, you can do anything after setting it. So these operations can see Property&amp; via `vm.property()`.
        })
    , m_other()
    , ...

This patch applies this to some of JSC::VM&apos;s fields, making definitions much cleaner by moving all messy parts into VM.cpp side and clean up VM.h&apos;s interface.

* Source/JavaScriptCore/runtime/HasOwnPropertyCache.h:
(JSC::HasOwnPropertyCache::Entry::offsetOfStructureID): Deleted.
(JSC::HasOwnPropertyCache::Entry::offsetOfImpl): Deleted.
(JSC::HasOwnPropertyCache::Entry::offsetOfResult): Deleted.
(JSC::HasOwnPropertyCache::operator delete): Deleted.
(JSC::HasOwnPropertyCache::create): Deleted.
(JSC::HasOwnPropertyCache::hash): Deleted.
(JSC::HasOwnPropertyCache::get): Deleted.
(JSC::HasOwnPropertyCache::tryAdd): Deleted.
(JSC::HasOwnPropertyCache::clear): Deleted.
(JSC::HasOwnPropertyCache::clearBuffer): Deleted.
(JSC::VM::ensureHasOwnPropertyCache): Deleted.
* Source/JavaScriptCore/runtime/LazyPropertyInlines.h:
(JSC::ElementType&gt;::initLater):
* Source/JavaScriptCore/runtime/VM.cpp:
(JSC::VM::VM):
(JSC::VM::~VM):
(JSC::VM::invalidateStructureChainIntegrity):
(JSC::VM::ensureWatchdog): Deleted.
(JSC::VM::ensureHeapProfiler): Deleted.
(JSC::VM::ensureShadowChicken): Deleted.
(JSC::VM::ensureMegamorphicCacheSlow): Deleted.
* Source/JavaScriptCore/runtime/VM.h:
(JSC::VM::watchdog):
(JSC::VM::ensureWatchdog):
(JSC::VM::heapProfiler):
(JSC::VM::ensureHeapProfiler):
(JSC::VM::hasOwnPropertyCache):
(JSC::VM::ensureHasOwnPropertyCache):
(JSC::VM::megamorphicCache):
(JSC::VM::ensureMegamorphicCache):
(JSC::VM::shadowChicken):
(JSC::VM::ensureShadowChicken):
(JSC::VM::heapProfiler const): Deleted.
* Source/WTF/WTF.xcodeproj/project.pbxproj:
* Source/WTF/wtf/CMakeLists.txt:
* Source/WTF/wtf/LazyRef.h: Added.
(WTF::LazyRef::~LazyRef):
(WTF::LazyRef::isInitialized const):
(WTF::LazyRef::get const):
(WTF::LazyRef::get):
(WTF::LazyRef::getIfExists const):
(WTF::LazyRef::getIfExists):
(WTF::LazyRef::initLater):
(WTF::LazyRef::set):
(WTF::LazyRef::callFunc):
* Source/WTF/wtf/LazyUniqueRef.h: Added.
(WTF::LazyUniqueRef::~LazyUniqueRef):
(WTF::LazyUniqueRef::isInitialized const):
(WTF::LazyUniqueRef::get const):
(WTF::LazyUniqueRef::get):
(WTF::LazyUniqueRef::getIfExists const):
(WTF::LazyUniqueRef::getIfExists):
(WTF::LazyUniqueRef::initLater):
(WTF::LazyUniqueRef::set):
(WTF::LazyUniqueRef::callFunc):

Canonical link: <a href="https://commits.webkit.org/273287@main">https://commits.webkit.org/273287@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/24267b3aad7acdf690ab449e87f04a39e0cde3f3

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/34992 "2 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/13877 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/37064 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/37736 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/31603 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/36150 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/16265 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/10955 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/30546 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/35534 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/11759 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/31202 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/10310 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/10365 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/31291 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/38988 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/29721 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/31823 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/31620 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/36389 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/34872 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/10469 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/8389 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/34375 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/12271 "Built successfully") | | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/41496 "Built successfully") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/11002 "Built successfully") | | [❌ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/8629 "Found 231 new JSC stress test failures: microbenchmarks/array-from-derived-object-func.js.bytecode-cache, microbenchmarks/array-from-derived-object-func.js.default, microbenchmarks/array-from-derived-object-func.js.dfg-eager, microbenchmarks/array-from-derived-object-func.js.dfg-eager-no-cjit-validate, microbenchmarks/array-from-derived-object-func.js.eager-jettison-no-cjit, microbenchmarks/array-from-derived-object-func.js.no-cjit-collect-continuously, microbenchmarks/array-from-derived-object-func.js.no-llint, microbenchmarks/array-from-object-func.js.bytecode-cache, microbenchmarks/array-from-object-func.js.default, microbenchmarks/array-from-object-func.js.dfg-eager ... (failure)") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/4498 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/11334 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

rniwa reviewed (approved):
Maybe we want to add tests for these?

Constellation committed:
[WTF] Add LazyRef & LazyUniqueRef
https://bugs.webkit.org/show_bug.cgi?id=267830
rdar://121328458

Reviewed by Ryosuke Niwa.

This patch adds LazyRef and LazyUniqueRef. This is similar to LazyProperty in JSC.
We can set *stateless* lambda in the constructor side of the owner object so that it offers clean interface for lazy initialization,
which does not mess up the owner object's interface.
For example,

    LazyUniqueRef<VM, Property> m_property;
    Property& property() { return m_property.get(*this); }

First parameter is owner, which needs to be passed to `get`. And then, this `get` will automatically initialize if it is not initialized.
And the initialization code can be freely customized, but you do not need to define it as a VM's member! Instead, you can define it as a lambda inside VM.cpp.

    // This lambda needs to be stateless. So it must not capture anything (otherwise, it hits crash because of RELEASE_ASSERT anyway).
    m_property.initLater(
        [](VM& vm, auto& ref) {
            // You can do whatever. Even you can invoke the other property which can be further lazily initialized (so, it implicitly creates dependency graph of initialization).
            ref.set(Property::create());
            // And further, you can do anything after setting it. So these operations can see Property& via `vm.property()`.
        });

Or you can set lambda in constructor, so like, VM's constructor list,

    : m_property(
        [](VM& vm, auto& ref) {
            // You can do whatever. Even you can invoke the other property which can be further lazily initialized (so, it implicitly creates dependency graph of initialization).
            ref.set(Property::create());
            // And further, you can do anything after setting it. So these operations can see Property& via `vm.property()`.
        })
    , m_other()
    , ...

This patch applies this to some of JSC::VM's fields, making definitions much cleaner by moving all messy parts into VM.cpp side and clean up VM.h's interface.

* Source/JavaScriptCore/runtime/HasOwnPropertyCache.h:
(JSC::HasOwnPropertyCache::Entry::offsetOfStructureID): Deleted.
(JSC::HasOwnPropertyCache::Entry::offsetOfImpl): Deleted.
(JSC::HasOwnPropertyCache::Entry::offsetOfResult): Deleted.
(JSC::HasOwnPropertyCache::operator delete): Deleted.
(JSC::HasOwnPropertyCache::create): Deleted.
(JSC::HasOwnPropertyCache::hash): Deleted.
(JSC::HasOwnPropertyCache::get): Deleted.
(JSC::HasOwnPropertyCache::tryAdd): Deleted.
(JSC::HasOwnPropertyCache::clear): Deleted.
(JSC::HasOwnPropertyCache::clearBuffer): Deleted.
(JSC::VM::ensureHasOwnPropertyCache): Deleted.
* Source/JavaScriptCore/runtime/LazyPropertyInlines.h:
(JSC::ElementType>::initLater):
* Source/JavaScriptCore/runtime/VM.cpp:
(JSC::VM::VM):
(JSC::VM::~VM):
(JSC::VM::invalidateStructureChainIntegrity):
(JSC::VM::ensureWatchdog): Deleted.
(JSC::VM::ensureHeapProfiler): Deleted.
(JSC::VM::ensureShadowChicken): Deleted.
(JSC::VM::ensureMegamorphicCacheSlow): Deleted.
* Source/JavaScriptCore/runtime/VM.h:
(JSC::VM::watchdog):
(JSC::VM::ensureWatchdog):
(JSC::VM::heapProfiler):
(JSC::VM::ensureHeapProfiler):
(JSC::VM::hasOwnPropertyCache):
(JSC::VM::ensureHasOwnPropertyCache):
(JSC::VM::megamorphicCache):
(JSC::VM::ensureMegamorphicCache):
(JSC::VM::shadowChicken):
(JSC::VM::ensureShadowChicken):
(JSC::VM::heapProfiler const): Deleted.
* Source/WTF/WTF.xcodeproj/project.pbxproj:
* Source/WTF/wtf/CMakeLists.txt:
* Source/WTF/wtf/LazyRef.h: Added.
(WTF::LazyRef::~LazyRef):
(WTF::LazyRef::isInitialized const):
(WTF::LazyRef::get const):
(WTF::LazyRef::get):
(WTF::LazyRef::getIfExists const):
(WTF::LazyRef::getIfExists):
(WTF::LazyRef::initLater):
(WTF::LazyRef::set):
(WTF::LazyRef::callFunc):
* Source/WTF/wtf/LazyUniqueRef.h: Added.
(WTF::LazyUniqueRef::~LazyUniqueRef):
(WTF::LazyUniqueRef::isInitialized const):
(WTF::LazyUniqueRef::get const):
(WTF::LazyUniqueRef::get):
(WTF::LazyUniqueRef::getIfExists const):
(WTF::LazyUniqueRef::getIfExists):
(WTF::LazyUniqueRef::initLater):
(WTF::LazyUniqueRef::set):
(WTF::LazyUniqueRef::callFunc):

Canonical link: https://commits.webkit.org/273287@main
